Reduce the amount of time spent "waiting for an agent"

##### Issue Type

- Feature enhancement

##### Summary

My team has slowly been trying to improve our CD system and pipelines lately. One of the best ways we've found to do that is to split our longer running processes into different stages and jobs, which helps a lot with debugging when something goes wrong and tracking progress in general. 

However, as nice as this is from a usability point of view, each new job increases the runtime for that task for about 25-30 seconds. Slightly more if there is some load on the server, slightly less if there isn't. I haven't found anything online to indicate that this behavior is uncommon, if anything it seems like this is relatively fast from the other threads I've seen. While this may not sound like a lot of time, what used to be a 5-minute stage that we split into 5 1-minute stages now takes 7.5 minutes (50% longer) to run. The usability and maintenance benefits *are* huge, but they don't seem worth such a huge hit to runtime, especially for pipelines that may run a few times a day.

Increasing the number of agents and parallelizing things (splitting into jobs) has helped mitigate the issue somewhat, but there's still plenty of things that have to run sequentially, and we'd prefer not to have to decide between maintainability and speed, especially since the bottleneck seems arbitrary and unnecessary.

I did some looking into the logs and found a few concrete things that seem like they could maybe be improved (see below), but I'm no expert so I may be way off the mark. I also noticed that after upgrading from 20 to 22.3.0 things got better by about 10 seconds, which is an indication that either *someone* is looking into it or the base image just got better, but I'd say it's still slow enough to discourage us from using what we consider to be the better workflow. Case in point, we're now discussing merging everything back into a single stage wherever reasonable to avoid these time penalties.

##### Environment
We're using elastic agents with the [gocd-agent-ubuntu-22.04](https://github.com/gocd/docker-gocd-agent-ubuntu-22.04) image. The rest of the environment is:

```
* Version: "22.3.0 (15301-9d23ed19a9ea46eaf7f18bd16671ae0569871f53)"
* JAVA Version: 17.0.5
* OS: Linux 4.18.0-305.19.1.el8_4.x86_64
* Available Processors: 4
* Average System Load: 0.02
```

##### Log snippets

The only "suspicious" behavior I found was this weird time between getting a cookie and getting a response from the server, it's the last thing that happens before actually running the job commands and seems to be almost exactly 10 seconds every time:
```
jvm 1    | 2023-01-18 10:56:04,797 [stdout] - 2023-01-18 10:56:04,796 INFO  [main] DefaultPluginJarChangeListener:74 - Plugin load finished: /go/plugins/external/docker-registry-artifact-plugin-1.2.0-127.jar
jvm 1    | 2023-01-18 10:56:04,797 [stdout] - 2023-01-18 10:56:04,797 INFO  [scheduler-1] AgentHTTPClientController:129 - About to get cookie from the server.
jvm 1    | 2023-01-18 10:56:04,830 [stdout] - 2023-01-18 10:56:04,830 INFO  [scheduler-1] AgentHTTPClientController:132 - Got cookie: -redacted-
jvm 1    | 2023-01-18 10:56:15,067 [stdout] - 2023-01-18 10:56:15,067 INFO  [scheduler-2] HttpService:125 - Got back 200 from server
```
I'm not exactly sure what's happening between those two log lines, but I suspect it's something that could be fixed by someone who actually knows what they're doing. 

A lot of time (~10s) also seems to be spent downloading jar and zip files for agent plugins.

```
jvm 1    | 2023-01-18 10:55:52,766 INFO  [TouchLoopThread-1] Lockfile:103 - Using lock file: /go/.agent-bootstrapper.running
jvm 1    | 2023-01-18 10:55:53,281 INFO  [WrapperJarAppMain] ServerBinaryDownloader:114 - Downloading agent.jar
jvm 1    | 2023-01-18 10:55:53,392 INFO  [WrapperJarAppMain] ServerBinaryDownloader:121 - Got server response
jvm 1    | 2023-01-18 10:55:56,622 INFO  [WrapperJarAppMain] ServerBinaryDownloader:129 - Piped the stream to admin/agent
...
jvm 1    | 2023-01-18 10:56:04,797 [stdout] - 2023-01-18 10:56:04,796 INFO  [main] DefaultPluginJarChangeListener:74 - Plugin load finished: /go/plugins/external/docker-registry-artifact-plugin-1.2.0-127.jar
```

I suspect this is something that could be done periodically on the image itself to avoid having to update it every time a job runs, but again I may be wrong and am happy to be corrected by someone who actually knows the reasoning behind it.

Hopefully something can be done, because we would really like to take full advantage of all of GoCDs features and it seems silly to go back to having single long-running jobs after seeing all the advantages that multiple jobs can provide.

Hi Chad, thanks for your quick and very informative reply!

To be clear, I never considered getting the cookies to be the problem, since that part seems to run through just fine, it's just the last thing that happens before the 10 seconds of silence. Unfortunately the ChatGPT suggestions also didn't seem to help me ;)

Sorry I omitted most of the plugin loading messages from my original logs, there is an error in there which I thought (perhaps wrongly) might be a red herring as well:

```
jvm 1    | 2023-01-18 14:58:15,540 [stdout] - 2023-01-18 14:58:15,539 INFO  [main] ThreadPoolTaskScheduler:166 - Initializing ExecutorService 'scheduler'
jvm 1    | 2023-01-18 14:58:15,869 [stdout] - 2023-01-18 14:58:15,869 INFO  [main] DefaultPluginJarChangeListener:67 - Plugin load starting: /go/plugins/bundled/gocd-yum-repository-poller-plugin.jar
jvm 1    | 2023-01-18 14:58:15,988 [stdout] - 2023-01-18 14:58:15,988 INFO  [main] DefaultPluginJarChangeListener:74 - Plugin load finished: /go/plugins/bundled/gocd-yum-repository-poller-plugin.jar
jvm 1    | 2023-01-18 14:58:15,998 [stdout] - 2023-01-18 14:58:15,998 INFO  [main] DefaultPluginJarChangeListener:67 - Plugin load starting: /go/plugins/external/git-fb-poller-1.3.5.jar
jvm 1    | 2023-01-18 14:58:16,320 [stdout] - 2023-01-18 14:58:16,320 INFO  [main] DefaultPluginJarChangeListener:74 - Plugin load finished: /go/plugins/external/git-fb-poller-1.3.5.jar
jvm 1    | 2023-01-18 14:58:16,330 [stdout] - 2023-01-18 14:58:16,330 INFO  [main] DefaultPluginJarChangeListener:67 - Plugin load starting: /go/plugins/external/github-pr-poller-1.3.5.jar
jvm 1    | 2023-01-18 14:58:16,390 [stderr] - SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
jvm 1    | 2023-01-18 14:58:16,390 [stderr] - SLF4J: Defaulting to no-operation (NOP) logger implementation
jvm 1    | 2023-01-18 14:58:16,390 [stderr] - SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
jvm 1    | 2023-01-18 14:58:16,725 [stdout] - 2023-01-18 14:58:16,725 INFO  [scheduler-1] SslInfrastructureService:78 - [Agent Registration] Starting to register agent.
jvm 1    | 2023-01-18 14:58:16,725 [stdout] - 2023-01-18 14:58:16,725 INFO  [scheduler-1] SslInfrastructureService:88 - [Agent Registration] Fetching token from server.
jvm 1    | 2023-01-18 14:58:16,810 [stdout] - 2023-01-18 14:58:16,810 INFO  [scheduler-1] TokenRequester:56 - The server has generated token for the agent.
jvm 1    | 2023-01-18 14:58:16,810 [stdout] - 2023-01-18 14:58:16,810 INFO  [scheduler-1] SslInfrastructureService:91 - [Agent Registration] Got a token from server.
jvm 1    | 2023-01-18 14:58:16,835 [stdout] - 2023-01-18 14:58:16,835 INFO  [main] DefaultPluginJarChangeListener:74 - Plugin load finished: /go/plugins/external/github-pr-poller-1.3.5.jar
jvm 1    | 2023-01-18 14:58:16,841 [stdout] - 2023-01-18 14:58:16,841 INFO  [scheduler-1] SslInfrastructureService:170 - This agent is now approved by the server.
jvm 1    | 2023-01-18 14:58:16,842 [stdout] - 2023-01-18 14:58:16,841 INFO  [scheduler-1] SslInfrastructureService:119 - [Agent Registration] Retrieved registration from Go server.
jvm 1    | 2023-01-18 14:58:16,846 [stdout] - 2023-01-18 14:58:16,845 INFO  [main] DefaultPluginJarChangeListener:67 - Plugin load starting: /go/plugins/external/docker-registry-artifact-plugin-1.2.0-127.jar
jvm 1    | 2023-01-18 14:58:16,867 [stdout] - 2023-01-18 14:58:16,867 INFO  [scheduler-1] SslInfrastructureService:81 - [Agent Registration] Successfully registered agent.
jvm 1    | 2023-01-18 14:58:17,019 [stdout] - 2023-01-18 14:58:17,018  INFO [main] p.c.g.a.d.r.c.g.a.d.r.DockerRegistryArtifactPlugin:72 [plugin-cd.go.artifact.docker.registry] - Loading plugin Artifact plugin for docker[null].
jvm 1    | 2023-01-18 14:58:17,160 [stdout] - 2023-01-18 14:58:17,159 ERROR [scheduler-2] RemotingClient:162 - Server responded to action `ping` with: status[422 Unprocessable Entity], body[{
jvm 1    | 2023-01-18 14:58:17,160 [stdout] -   "message" : "Agent [Agent [185d037fb16d, 172.17.0.3, de964d9b-aa92-4c97-9328-e59d4a63c331, null]] has no cookie set; nested exception is com.thoughtworks.go.server.service.AgentNoCookieSetException: Agent [Agent [185d037fb16d, 172.17.0.3, de964d9b-aa92-4c97-9328-e59d4a63c331, null]] has no cookie set"
jvm 1    | 2023-01-18 14:58:17,160 [stdout] - }]
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 2023-01-18 14:58:17,160 ERROR [scheduler-2] AgentHTTPClientController:104 - Error occurred when agent tried to ping server:
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - java.lang.RuntimeException: org.apache.http.client.ClientProtocolException: The server returned status code 422. Possible reasons include:
jvm 1    | 2023-01-18 14:58:17,161 [stdout] -    - This agent has been deleted from the configuration
jvm 1    | 2023-01-18 14:58:17,161 [stdout] -    - This agent is pending approval
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at com.thoughtworks.go.agent.RemotingClient.post(RemotingClient.java:117)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at com.thoughtworks.go.agent.RemotingClient.ping(RemotingClient.java:73)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at com.thoughtworks.go.agent.AgentHTTPClientController.ping(AgentHTTPClientController.java:99)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at java.base/java.lang.Thread.run(Unknown Source)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - Caused by: org.apache.http.client.ClientProtocolException: The server returned status code 422. Possible reasons include:
jvm 1    | 2023-01-18 14:58:17,161 [stdout] -    - This agent has been deleted from the configuration
jvm 1    | 2023-01-18 14:58:17,161 [stdout] -    - This agent is pending approval
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at com.thoughtworks.go.agent.RemotingClient.validateResponse(RemotingClient.java:140)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	at com.thoughtworks.go.agent.RemotingClient.post(RemotingClient.java:113)
jvm 1    | 2023-01-18 14:58:17,161 [stdout] - 	... 14 common frames omitted
jvm 1    | 2023-01-18 14:58:18,063 [stdout] - 2023-01-18 14:58:18,063 INFO  [main] DefaultPluginJarChangeListener:74 - Plugin load finished: /go/plugins/external/docker-registry-artifact-plugin-1.2.0-127.jar
jvm 1    | 2023-01-18 14:58:18,063 [stdout] - 2023-01-18 14:58:18,063 INFO  [scheduler-1] AgentHTTPClientController:129 - About to get cookie from the server.
jvm 1    | 2023-01-18 14:58:18,096 [stdout] - 2023-01-18 14:58:18,096 INFO  [scheduler-1] AgentHTTPClientController:132 - Got cookie: 57682c74-ee25-4a29-9698-c64d314df0ed
jvm 1    | 2023-01-18 14:58:28,341 [stdout] - 2023-01-18 14:58:28,340 INFO  [scheduler-2] HttpService:125 - Got back 200 from server
jvm 1    | 2023-01-18 14:58:28,341 [stdout] - 2023-01-18 14:58:28,341 INFO  [scheduler-2] DefaultGoPublisher:89 - Agent [185d037fb16d, 172.17.0.3, de964d9b-aa92-4c97-9328-e59d4a63c331] is reporting status [Preparing] to Go Server for Build [docker-rb-vision-develop/497/Build-Deploy/1/make-deploy/7013]
jvm 1    | 2023-01-18 14:58:31,670 [stdout] - 2023-01-18 14:58:31,669 INFO  [scheduler-2] HttpService:125 - Got back 200 from server
jvm 1    | 2023-01-18 14:58:31,670 [stdout] - 2023-01-18 14:58:31,670 INFO  [scheduler-2] DefaultGoPublisher:89 - Agent [185d037fb16d, 172.17.0.3, de964d9b-aa92-4c97-9328-e59d4a63c331] is reporting status [Building] to Go Server for Build [docker-rb-vision-develop/497/Build-Deploy/1/make-deploy/7013]
jvm 1    | 2023-01-18 14:58:31,755 [stdout] - 2023-01-18 14:58:31,755 INFO  [scheduler-2] CommandLine:316 - Running command: make deploy-develop
```

As you can see, the last "load finished" message is about 10 seconds before the "got 200 back" message, which is what lead me to believe that it's not an issue with loading plugins. However, I still have no idea what is happening between those two lines in the logs. For the sake of completeness, it looks like the only plugins we're loading are:
 * `yum-repo-poller`, which is referenced in your link
 * `docker-registry-artifact`
 * `git-fb-poller`
 * `github-pr-poller`

The total loading time for all the plugins is (relatively) low, at ~2.5s, so again I don't think that's the crux of the problem. I could very well be misunderstanding the nature of how plugins are loaded, though, feel free to correct me if I'm wrong.

I do think our network on our server, our agent, or both is quite slow seeing as it takes so long to download everything. However, my point still stands in that this is not something that I think should happen every time a job runs. I may misunderstand how plugins are supposed to work, but couldn't they be kept up to date on the image itself?

Not sure what you mean by the type of agents we're using, but I'll try to answer to the best of my ability... We're using the [docker elastic agents plugin](https://github.com/gocd-contrib/docker-elastic-agents-plugin) on our server, and almost all of our pipelines run on the aforementioned ubuntu 22 docker agent image.  As far as I can tell this is the same version as the server, not 100% sure though. We have a single separate server where all the agents run, which is sufficient for us since only a few (< 5) are running at any given time.

I've also attached full logs until agent picks up the first job, as requested: [gocd-logs.txt](https://github.com/gocd/gocd/files/10448523/gocd-logs.txt)

Side note, and sorry if this is a silly question, but is it possible to use the same elastic agent for multiple jobs/stages? I didn't come accross this at all during my reading of the documentation or other threads, but it would pretty much eliminate the downsides since we wouldn't have to wait for the agent to restart every time.

--- a/agent/src/main/java/com/thoughtworks/go/agent/AgentController.java
+++ b/agent/src/main/java/com/thoughtworks/go/agent/AgentController.java
@@ -25,6 +25,7 @@ import com.thoughtworks.go.domain.AgentRuntimeStatus;
 import com.thoughtworks.go.domain.AgentStatus;
 import com.thoughtworks.go.plugin.infra.PluginManager;
 import com.thoughtworks.go.plugin.infra.PluginManagerReference;
+import com.thoughtworks.go.plugin.infra.monitor.PluginJarLocationMonitor;
 import com.thoughtworks.go.remote.AgentIdentifier;
 import com.thoughtworks.go.server.service.AgentRuntimeInfo;
 import com.thoughtworks.go.server.service.ElasticAgentRuntimeInfo;
@@ -51,6 +52,7 @@ public abstract class AgentController {
     private final SubprocessLogger subprocessLogger;
     private final AgentUpgradeService agentUpgradeService;
     private final AgentHealthHolder agentHealthHolder;
+    private final PluginJarLocationMonitor pluginJarLocationMonitor;
     private final String hostName;
     private final String ipAddress;
 
@@ -60,12 +62,14 @@ public abstract class AgentController {
                            PluginManager pluginManager,
                            SubprocessLogger subprocessLogger,
                            AgentUpgradeService agentUpgradeService,
-                           AgentHealthHolder agentHealthHolder) {
+                           AgentHealthHolder agentHealthHolder,
+                           PluginJarLocationMonitor pluginJarLocationMonitor) {
         this.sslInfrastructureService = sslInfrastructureService;
         this.agentRegistry = agentRegistry;
         this.subprocessLogger = subprocessLogger;
         this.agentUpgradeService = agentUpgradeService;
         this.agentHealthHolder = agentHealthHolder;
+        this.pluginJarLocationMonitor = pluginJarLocationMonitor;
         PluginManagerReference.reference().setPluginManager(pluginManager);
         hostName = SystemUtil.getLocalhostNameOrRandomNameIfNotFound();
         ipAddress = SystemUtil.getClientIp(systemEnvironment.getServiceUrl());
@@ -75,28 +79,34 @@ public abstract class AgentController {
 
     public abstract void execute();
 
-    public final void loop() {
+    final WorkAttempt performWork() {
         try {
             LOG.debug("[Agent Loop] Trying to retrieve work.");
             agentUpgradeService.checkForUpgradeAndExtraProperties();
             sslInfrastructureService.registerIfNecessary(getAgentAutoRegistrationProperties());
-            work();
-            LOG.debug("[Agent Loop] Successfully retrieved work.");
+
+            if (pluginJarLocationMonitor.hasRunAtLeastOnce()) {
+                return tryDoWork();
+            } else {
+                LOG.debug("[Agent Loop] PluginLocationMonitor has not yet run. Not retrieving work since plugins may not be initialized.");
+                return WorkAttempt.FAILED;
+            }
+
         } catch (Exception e) {
             if (isCausedBySecurity(e)) {
                 handleSecurityException(e);
             } else {
                 LOG.error("[Agent Loop] Error occurred during loop: ", e);
             }
+            return WorkAttempt.FAILED;
         }
-
     }
 
     private void handleSecurityException(Exception e) {
         LOG.error("There has been a problem with one of GoCD's TLS certificates. This can be caused by a man-in-the-middle attack, or a change to the HTTPS certificates of the GoCD Server. Review the agent TLS trust settings and any mutual TLS configuration of the agent.", e);
     }
 
-    protected abstract void work();
+    protected abstract WorkAttempt tryDoWork();
 
     protected AgentAutoRegistrationProperties getAgentAutoRegistrationProperties() {
         return agentAutoRegistrationProperties;

--- a/agent/src/main/java/com/thoughtworks/go/agent/AgentHTTPClientController.java
+++ b/agent/src/main/java/com/thoughtworks/go/agent/AgentHTTPClientController.java
@@ -51,7 +51,6 @@ public class AgentHTTPClientController extends AgentController {
     private final SslInfrastructureService sslInfrastructureService;
     private final ArtifactExtension artifactExtension;
     private final PluginRequestProcessorRegistry pluginRequestProcessorRegistry;
-    private final PluginJarLocationMonitor pluginJarLocationMonitor;
 
     private final PackageRepositoryExtension packageRepositoryExtension;
     private final SCMExtension scmExtension;
@@ -75,7 +74,7 @@ public class AgentHTTPClientController extends AgentController {
                                      PluginRequestProcessorRegistry pluginRequestProcessorRegistry,
                                      AgentHealthHolder agentHealthHolder,
                                      PluginJarLocationMonitor pluginJarLocationMonitor) {
-        super(sslInfrastructureService, systemEnvironment, agentRegistry, pluginManager, subprocessLogger, agentUpgradeService, agentHealthHolder);
+        super(sslInfrastructureService, systemEnvironment, agentRegistry, pluginManager, subprocessLogger, agentUpgradeService, agentHealthHolder, pluginJarLocationMonitor);
         this.client = client;
         this.packageRepositoryExtension = packageRepositoryExtension;
         this.scmExtension = scmExtension;
@@ -84,7 +83,6 @@ public class AgentHTTPClientController extends AgentController {
         this.sslInfrastructureService = sslInfrastructureService;
         this.artifactExtension = artifactExtension;
         this.pluginRequestProcessorRegistry = pluginRequestProcessorRegistry;
-        this.pluginJarLocationMonitor = pluginJarLocationMonitor;
     }
 
     @Override
@@ -113,15 +111,9 @@ public class AgentHTTPClientController extends AgentController {
     }
 
     @Override
-    public void work() {
-        LOG.debug("[Agent Loop] Trying to retrieve work.");
-        if (pluginJarLocationMonitor.hasRunAtLeastOnce()) {
-            retrieveCookieIfNecessary();
-            retrieveWork();
-            LOG.debug("[Agent Loop] Successfully retrieved work.");
-        } else {
-            LOG.debug("[Agent Loop] PluginLocationMonitor has not yet run. Not retrieving work since plugins may not be initialized.");
-        }
+    public WorkAttempt tryDoWork() {
+        retrieveCookieIfNecessary();
+        return doWork();
     }
 
     private void retrieveCookieIfNecessary() {
@@ -133,22 +125,24 @@ public class AgentHTTPClientController extends AgentController {
         }
     }
 
-    void retrieveWork() {
+    private WorkAttempt doWork() {
         AgentIdentifier agentIdentifier = agentIdentifier();
         LOG.debug("[Agent Loop] {} is checking for work from Go", agentIdentifier);
-        Work work;
         try {
             getAgentRuntimeInfo().idle();
-            work = client.getWork(getAgentRuntimeInfo());
+            Work work = client.getWork(getAgentRuntimeInfo());
             if (!(work instanceof NoWork)) {
                 LOG.debug("[Agent Loop] Got work from server: [{}]", work.description());
             }
             runner = new JobRunner();
             final AgentWorkContext agentWorkContext = new AgentWorkContext(agentIdentifier, client, manipulator, getAgentRuntimeInfo(), packageRepositoryExtension, scmExtension, taskExtension, artifactExtension, pluginRequestProcessorRegistry);
             runner.run(work, agentWorkContext);
+            LOG.debug("[Agent Loop] Successfully executed work.");
+            return WorkAttempt.fromWork(work);
         } catch (UnregisteredAgentException e) {
             LOG.warn("[Agent Loop] Agent is not registered. [{}] Registering with server on next iteration.", e.getMessage());
             sslInfrastructureService.createSslInfrastructure();
+            return WorkAttempt.FAILED;
         } finally {
             getAgentRuntimeInfo().idle();
         }

--- a/agent/src/main/java/com/thoughtworks/go/agent/AgentWorkRetrievalScheduler.java
+++ b/agent/src/main/java/com/thoughtworks/go/agent/AgentWorkRetrievalScheduler.java
@@ -1,0 +1,75 @@
+/*
+ * Copyright 2023 Thoughtworks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.agent;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.support.PeriodicTrigger;
+import org.springframework.stereotype.Component;
+import org.springframework.util.backoff.BackOff;
+import org.springframework.util.backoff.BackOffExecution;
+
+import javax.annotation.PostConstruct;
+import java.util.concurrent.TimeUnit;
+
+@Component
+public class AgentWorkRetrievalScheduler implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(AgentWorkRetrievalScheduler.class);
+
+    private final AgentController controller;
+    private final BackOff backoffStrategy;
+    private final TaskScheduler scheduler;
+
+    @Autowired
+    public AgentWorkRetrievalScheduler(AgentController controller, BackOff backoffStrategy, TaskScheduler scheduler) {
+        this.controller = controller;
+        this.backoffStrategy = backoffStrategy;
+        this.scheduler = scheduler;
+    }
+
+    @PostConstruct
+    public void schedule() {
+        // Schedule constantly, to ensure any unexpected exceptions in the loop don't cause a zombie agent
+        scheduler.schedule(this, new PeriodicTrigger(1, TimeUnit.MILLISECONDS));
+    }
+
+    @Override
+    public void run() {
+        BackOffExecution backOffExecution = backoffStrategy.start();
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                long nextBackOffMillis = backOffExecution.nextBackOff();
+                LOG.debug("[Agent Loop] Waiting {} ms before retrieving next work.", nextBackOffMillis);
+                waitFor(nextBackOffMillis);
+                WorkAttempt result = controller.performWork();
+                LOG.debug("[Agent Loop] Work attempted was {}", result);
+                if (result.shouldResetDelay()) {
+                    backOffExecution = backoffStrategy.start();
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    void waitFor(long waitMillis) throws InterruptedException {
+        Thread.sleep(waitMillis);
+    }
+
+}

--- a/agent/src/main/java/com/thoughtworks/go/agent/WorkAttempt.java
+++ b/agent/src/main/java/com/thoughtworks/go/agent/WorkAttempt.java
@@ -1,0 +1,41 @@
+/*
+ * Copyright 2023 Thoughtworks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.agent;
+
+import com.thoughtworks.go.remote.work.*;
+
+import java.util.Map;
+
+enum WorkAttempt {
+    OK, NOTHING_TO_DO, FAILED;
+
+    private static final Map<Class<? extends Work>, WorkAttempt> WORK_TO_RESULT = Map.of(
+        BuildWork.class, OK,
+        NoWork.class, NOTHING_TO_DO,
+        DeniedAgentWork.class, NOTHING_TO_DO,
+        UnregisteredAgentWork.class, FAILED
+    );
+
+    public static WorkAttempt fromWork(Work work) {
+        return WORK_TO_RESULT.getOrDefault(work.getClass(), OK);
+    }
+
+    boolean shouldResetDelay() {
+        // Reset backoff delays once we have executed real work successfully
+        return OK.equals(this);
+    }
+}

--- a/agent/src/main/resources/agent.properties
+++ b/agent/src/main/resources/agent.properties
@@ -20,3 +20,4 @@ agent.ping.interval=5000
 agent.ping.delay=1500
 agent.get.work.interval=10000
 agent.get.work.delay=1000
+agent.get.work.backOffMultiplier=1.5

--- a/agent/src/main/resources/applicationContext.xml
+++ b/agent/src/main/resources/applicationContext.xml
@@ -70,12 +70,20 @@
   <task:scheduler id="scheduler" pool-size="3"/>
 
   <task:scheduled-tasks scheduler="scheduler">
-    <task:scheduled ref="agentController" method="loop"
-                    initial-delay="${agent.get.work.delay}" fixed-delay="${agent.get.work.interval}"/>
     <task:scheduled ref="agentController" method="ping"
                     initial-delay="${agent.ping.delay}" fixed-delay="${agent.ping.interval}"/>
     <task:scheduled ref="agentController" method="execute"
                     initial-delay="${agent.instruction.delay}" fixed-delay="${agent.instruction.interval}"/>
   </task:scheduled-tasks>
+
+  <bean id="agentWorkScheduler" class="com.thoughtworks.go.agent.AgentWorkRetrievalScheduler">
+    <constructor-arg>
+      <bean class="org.springframework.util.backoff.ExponentialBackOff">
+        <property name="initialInterval" value="${agent.get.work.delay}"/>
+        <property name="maxInterval" value="${agent.get.work.interval}"/>
+        <property name="multiplier" value="${agent.get.work.backOffMultiplier}"/>
+      </bean>
+    </constructor-arg>
+  </bean>
 
 </beans>

--- a/agent/src/main/resources/config/agent-logback.xml
+++ b/agent/src/main/resources/config/agent-logback.xml
@@ -45,6 +45,7 @@
     <appender-ref ref="${gocd.agent.logback.root.appender:-FileAppender}"/>
   </root>
 
+  <logger name="com.thoughtworks.go.agent" level="INFO"/>
   <logger name="org.apache.http.wire" level="ERROR"/>
 
   <!-- make sure this is the last line in the config -->

--- a/agent/src/test/java/com/thoughtworks/go/agent/AgentControllerTest.java
+++ b/agent/src/test/java/com/thoughtworks/go/agent/AgentControllerTest.java
@@ -20,9 +20,11 @@ import com.thoughtworks.go.agent.service.SslInfrastructureService;
 import com.thoughtworks.go.agent.statusapi.AgentHealthHolder;
 import com.thoughtworks.go.config.AgentRegistry;
 import com.thoughtworks.go.plugin.infra.PluginManager;
+import com.thoughtworks.go.plugin.infra.monitor.PluginJarLocationMonitor;
 import com.thoughtworks.go.util.SubprocessLogger;
 import com.thoughtworks.go.util.SystemEnvironment;
 import com.thoughtworks.go.util.TestingClock;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.InOrder;
@@ -34,7 +36,7 @@ import java.security.GeneralSecurityException;
 import java.util.Date;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.*;
 
 @ExtendWith(MockitoExtension.class)
 public class AgentControllerTest {
@@ -48,36 +50,48 @@ public class AgentControllerTest {
     private AgentUpgradeService agentUpgradeService;
     @Mock
     private PluginManager pluginManager;
-    private AgentController agentController;
-
     @Mock
     private AgentRegistry agentRegistry;
-    private TestingClock clock = new TestingClock();
+    @Mock
+    private PluginJarLocationMonitor pluginJarLocationMonitor;
+    private AgentController agentController;
+
+    private final TestingClock clock = new TestingClock();
     private final int pingInterval = 5000;
-    private AgentHealthHolder agentHealthHolder = new AgentHealthHolder(clock, pingInterval);
+    private final AgentHealthHolder agentHealthHolder = new AgentHealthHolder(clock, pingInterval);
+
+    @BeforeEach
+    public void setUp() {
+        agentController = createAgentController();
+    }
 
     @Test
     void shouldReturnTrueIfCausedBySecurity() {
         Exception exception = new Exception(new RuntimeException(new GeneralSecurityException()));
-
-        agentController = createAgentController();
         assertThat(agentController.isCausedBySecurity(exception)).isTrue();
     }
 
     @Test
     void shouldReturnFalseIfNotCausedBySecurity() {
         Exception exception = new Exception(new IOException());
-        agentController = createAgentController();
         assertThat(agentController.isCausedBySecurity(exception)).isFalse();
     }
 
     @Test
     void shouldUpgradeAgentBeforeAgentRegistration() throws Exception {
-        agentController = createAgentController();
-        InOrder inOrder = inOrder(agentUpgradeService, sslInfrastructureService);
-        agentController.loop();
+        when(pluginJarLocationMonitor.hasRunAtLeastOnce()).thenReturn(true);
+        assertThat(agentController.performWork()).isEqualTo(WorkAttempt.OK);
+        InOrder inOrder = inOrder(agentUpgradeService, sslInfrastructureService, pluginJarLocationMonitor);
         inOrder.verify(agentUpgradeService).checkForUpgradeAndExtraProperties();
         inOrder.verify(sslInfrastructureService).registerIfNecessary(agentController.getAgentAutoRegistrationProperties());
+        inOrder.verify(pluginJarLocationMonitor).hasRunAtLeastOnce();
+    }
+
+    @Test
+    void shouldNotTryWorkIfPluginMonitorHasNotRun() {
+        when(pluginJarLocationMonitor.hasRunAtLeastOnce()).thenReturn(false);
+        assertThat(agentController.performWork()).isEqualTo(WorkAttempt.FAILED);
+        verify(pluginJarLocationMonitor).hasRunAtLeastOnce();
     }
 
     @Test
@@ -85,7 +99,6 @@ public class AgentControllerTest {
         // initial time
         Date now = new Date(42);
         clock.setTime(now);
-        agentController = createAgentController();
         agentController.pingSuccess();
 
         assertThat(agentHealthHolder.hasLostContact()).isFalse();
@@ -97,20 +110,18 @@ public class AgentControllerTest {
 
     private AgentController createAgentController() {
         return new AgentController(sslInfrastructureService, systemEnvironment, agentRegistry, pluginManager,
-                subprocessLogger, agentUpgradeService, agentHealthHolder) {
+            subprocessLogger, agentUpgradeService, agentHealthHolder, pluginJarLocationMonitor) {
             @Override
             public void ping() {
-
             }
 
             @Override
             public void execute() {
-
             }
 
             @Override
-            protected void work() {
-
+            protected WorkAttempt tryDoWork() {
+                return WorkAttempt.OK;
             }
         };
     }

--- a/agent/src/test/java/com/thoughtworks/go/agent/AgentHTTPClientControllerTest.java
+++ b/agent/src/test/java/com/thoughtworks/go/agent/AgentHTTPClientControllerTest.java
@@ -19,6 +19,7 @@ import com.thoughtworks.go.agent.service.AgentUpgradeService;
 import com.thoughtworks.go.agent.service.SslInfrastructureService;
 import com.thoughtworks.go.config.AgentRegistry;
 import com.thoughtworks.go.config.GuidService;
+import com.thoughtworks.go.domain.exception.UnregisteredAgentException;
 import com.thoughtworks.go.plugin.access.artifact.ArtifactExtension;
 import com.thoughtworks.go.plugin.access.packagematerial.PackageRepositoryExtension;
 import com.thoughtworks.go.plugin.access.pluggabletask.TaskExtension;
@@ -28,6 +29,8 @@ import com.thoughtworks.go.plugin.infra.PluginManagerReference;
 import com.thoughtworks.go.plugin.infra.monitor.PluginJarLocationMonitor;
 import com.thoughtworks.go.publishers.GoArtifactsManipulator;
 import com.thoughtworks.go.remote.work.AgentWorkContext;
+import com.thoughtworks.go.remote.work.DeniedAgentWork;
+import com.thoughtworks.go.remote.work.NoWork;
 import com.thoughtworks.go.remote.work.Work;
 import com.thoughtworks.go.server.service.AgentRuntimeInfo;
 import com.thoughtworks.go.util.SubprocessLogger;
@@ -38,8 +41,6 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
-
-import java.io.IOException;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.*;
@@ -89,31 +90,16 @@ public class AgentHTTPClientControllerTest {
     }
 
     @Test
-    void shouldRetrieveWorkFromServerAndDoIt() throws Exception {
-        when(loopServer.getWork(any(AgentRuntimeInfo.class))).thenReturn(work);
-        when(agentRegistry.uuid()).thenReturn(agentUuid);
-        when(pluginJarLocationMonitor.hasRunAtLeastOnce()).thenReturn(true);
-        agentController = createAgentController();
-        agentController.init();
+    void shouldRetrieveWorkFromServerAndDoIt() {
+        prepareForWork();
         agentController.ping();
-        agentController.work();
+        assertThat(agentController.tryDoWork()).isEqualTo(WorkAttempt.OK);
         verify(work).doWork(any(EnvironmentVariableContext.class), any(AgentWorkContext.class));
         verify(sslInfrastructureService).createSslInfrastructure();
     }
 
     @Test
-    void shouldNotRetrieveWorkIfPluginMonitorHasNotRun() throws IOException {
-        when(agentRegistry.uuid()).thenReturn(agentUuid);
-        when(pluginJarLocationMonitor.hasRunAtLeastOnce()).thenReturn(false);
-        agentController = createAgentController();
-        agentController.init();
-        agentController.ping();
-        agentController.work();
-        verifyNoMoreInteractions(work);
-    }
-
-    @Test
-    void shouldRetrieveCookieIfNotPresent() throws Exception {
+    void shouldRetrieveCookieIfNotPresent() {
         agentController = createAgentController();
         agentController.init();
 
@@ -122,34 +108,57 @@ public class AgentHTTPClientControllerTest {
         when(loopServer.getWork(agentController.getAgentRuntimeInfo())).thenReturn(work);
         when(agentRegistry.uuid()).thenReturn(agentUuid);
         when(pluginJarLocationMonitor.hasRunAtLeastOnce()).thenReturn(true);
-        agentController.loop();
-        verify(work).doWork(any(EnvironmentVariableContext.class), any(AgentWorkContext.class));
+        assertThat(agentController.performWork()).isEqualTo(WorkAttempt.OK);
+        verify(work).doWork(any(), any());
     }
 
     @Test
-    void shouldNotTellServerWorkIsCompletedWhenThereIsNoWork() throws Exception {
-        when(loopServer.getWork(any(AgentRuntimeInfo.class))).thenReturn(work);
-        when(agentRegistry.uuid()).thenReturn(agentUuid);
-        agentController = createAgentController();
-        agentController.init();
-        agentController.retrieveWork();
-        verify(work).doWork(any(EnvironmentVariableContext.class), any(AgentWorkContext.class));
+    void shouldNotTellServerWorkIsCompletedWhenThereIsNoWork() {
+        prepareForWork();
+        assertThat(agentController.tryDoWork()).isEqualTo(WorkAttempt.OK);
+        verify(work).doWork(any(), any());
         verify(sslInfrastructureService).createSslInfrastructure();
     }
 
     @Test
-    void shouldRegisterSubprocessLoggerAtExit() throws Exception {
-        SslInfrastructureService sslInfrastructureService = mock(SslInfrastructureService.class);
-        AgentRegistry agentRegistry = mock(AgentRegistry.class);
-        agentController = new AgentHTTPClientController(loopServer, artifactsManipulator, sslInfrastructureService,
-                agentRegistry, agentUpgradeService, subprocessLogger, systemEnvironment, pluginManager,
-                packageRepositoryExtension, scmExtension, taskExtension, artifactExtension, null, null, pluginJarLocationMonitor);
+    void workStatusShouldBeFailedWhenUnregisteredAgentExceptionThrown() {
+        prepareForWork();
+
+        doThrow(UnregisteredAgentException.class).when(work).doWork(any(), any());
+
+        assertThat(agentController.tryDoWork()).isEqualTo(WorkAttempt.FAILED);
+    }
+
+    @Test
+    void workStatusShouldDeriveFromWorkTypeForNoWork() {
+        work = mock(NoWork.class);
+        prepareForWork();
+        assertThat(agentController.tryDoWork()).isEqualTo(WorkAttempt.NOTHING_TO_DO);
+    }
+
+    @Test
+    void workStatusShouldDeriveFromWorkTypeForDeniedWork() {
+        work = mock(DeniedAgentWork.class);
+        prepareForWork();
+        assertThat(agentController.tryDoWork()).isEqualTo(WorkAttempt.NOTHING_TO_DO);
+    }
+
+    private void prepareForWork() {
+        when(loopServer.getWork(any(AgentRuntimeInfo.class))).thenReturn(work);
+        when(agentRegistry.uuid()).thenReturn(agentUuid);
+        agentController = createAgentController();
+        agentController.init();
+    }
+
+    @Test
+    void shouldRegisterSubprocessLoggerAtExit() {
+        agentController = createAgentController();
         agentController.init();
         verify(subprocessLogger).registerAsExitHook("Following processes were alive at shutdown: ");
     }
 
     @Test
-    void shouldNotPingIfNotRegisteredYet() throws Exception {
+    void shouldNotPingIfNotRegisteredYet() {
         when(agentRegistry.uuid()).thenReturn(agentUuid);
         when(sslInfrastructureService.isRegistered()).thenReturn(false);
 
@@ -160,7 +169,7 @@ public class AgentHTTPClientControllerTest {
     }
 
     @Test
-    void shouldPingIfAfterRegistered() throws Exception {
+    void shouldPingIfAfterRegistered() {
         when(agentRegistry.uuid()).thenReturn(agentUuid);
         when(sslInfrastructureService.isRegistered()).thenReturn(true);
         agentController = createAgentController();
@@ -173,17 +182,17 @@ public class AgentHTTPClientControllerTest {
     private AgentHTTPClientController createAgentController() {
 
         return new AgentHTTPClientController(
-                loopServer,
-                artifactsManipulator,
-                sslInfrastructureService,
-                agentRegistry,
-                agentUpgradeService,
-                subprocessLogger,
-                systemEnvironment,
-                pluginManager,
-                packageRepositoryExtension,
-                scmExtension,
-                taskExtension,
-                artifactExtension, null, null, pluginJarLocationMonitor);
+            loopServer,
+            artifactsManipulator,
+            sslInfrastructureService,
+            agentRegistry,
+            agentUpgradeService,
+            subprocessLogger,
+            systemEnvironment,
+            pluginManager,
+            packageRepositoryExtension,
+            scmExtension,
+            taskExtension,
+            artifactExtension, null, null, pluginJarLocationMonitor);
     }
 }

--- a/agent/src/test/java/com/thoughtworks/go/agent/AgentWorkRetrievalSchedulerTest.java
+++ b/agent/src/test/java/com/thoughtworks/go/agent/AgentWorkRetrievalSchedulerTest.java
@@ -1,0 +1,142 @@
+/*
+ * Copyright 2023 Thoughtworks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.agent;
+
+import ch.qos.logback.classic.Level;
+import com.thoughtworks.go.util.LogFixture;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.backoff.ExponentialBackOff;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+@MockitoSettings
+class AgentWorkRetrievalSchedulerTest {
+
+    @Mock
+    private AgentController controller;
+
+    @Mock
+    private TaskScheduler taskScheduler;
+
+    @Test
+    void shouldLoopForWorkWithExponentialBackoffs() throws InterruptedException {
+        AgentWorkRetrievalScheduler scheduler = createSchedulerForIterations(exponentialBackOffTwoToTen(), 10);
+
+        when(controller.performWork())
+            .thenReturn(WorkAttempt.FAILED)
+            .thenReturn(WorkAttempt.FAILED)
+            .thenReturn(WorkAttempt.NOTHING_TO_DO)
+            .thenReturn(WorkAttempt.NOTHING_TO_DO)
+            .thenReturn(WorkAttempt.OK)
+            .thenReturn(WorkAttempt.OK)
+            .thenReturn(WorkAttempt.NOTHING_TO_DO);
+
+        try (LogFixture logging = LogFixture.logFixtureFor(AgentWorkRetrievalScheduler.class, Level.DEBUG)) {
+            Thread runner = new Thread(scheduler);
+
+            runner.start();
+            runner.join();
+
+            verify(controller, times(10)).performWork();
+
+            assertThat(logging.getRawMessages().stream().filter(x -> x.startsWith("[Agent Loop] Waiting")))
+                .containsExactly(
+                    "[Agent Loop] Waiting 2 ms before retrieving next work.", // Initial delay
+                    "[Agent Loop] Waiting 4 ms before retrieving next work.", // Failed 1
+                    "[Agent Loop] Waiting 8 ms before retrieving next work.", // Failed 2
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Nothing to do 1 (at maximum)
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Nothing to do 2 (at maximum)
+                    "[Agent Loop] Waiting 2 ms before retrieving next work.", // After OK - reset
+                    "[Agent Loop] Waiting 2 ms before retrieving next work.", // After OK - reset
+                    "[Agent Loop] Waiting 4 ms before retrieving next work.", // Nothing to do
+                    "[Agent Loop] Waiting 8 ms before retrieving next work.", // Nothing to do
+                    "[Agent Loop] Waiting 10 ms before retrieving next work." // Nothing to do
+                );
+        }
+    }
+
+    @Test
+    void shouldLoopForWorkWithInstantMaxBackoff() throws InterruptedException {
+        AgentWorkRetrievalScheduler scheduler = createSchedulerForIterations(instantBackOffTwoToTen(), 9);
+
+        when(controller.performWork())
+            .thenReturn(WorkAttempt.FAILED)
+            .thenReturn(WorkAttempt.FAILED)
+            .thenReturn(WorkAttempt.NOTHING_TO_DO)
+            .thenReturn(WorkAttempt.NOTHING_TO_DO)
+            .thenReturn(WorkAttempt.OK)
+            .thenReturn(WorkAttempt.OK)
+            .thenReturn(WorkAttempt.NOTHING_TO_DO);
+
+        try (LogFixture logging = LogFixture.logFixtureFor(AgentWorkRetrievalScheduler.class, Level.DEBUG)) {
+            Thread runner = new Thread(scheduler);
+
+            runner.start();
+            runner.join();
+
+            verify(controller, times(9)).performWork();
+
+            assertThat(logging.getRawMessages().stream().filter(x -> x.startsWith("[Agent Loop] Waiting")))
+                .containsExactly(
+                    "[Agent Loop] Waiting 2 ms before retrieving next work.", // Initial delay
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Failed 1 (at maximum)
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Failed 2 (at maximum)
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Nothing to do 1 (at maximum)
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Nothing to do 2 (at maximum)
+                    "[Agent Loop] Waiting 2 ms before retrieving next work.", // After OK - reset
+                    "[Agent Loop] Waiting 2 ms before retrieving next work.", // After OK - reset
+                    "[Agent Loop] Waiting 10 ms before retrieving next work.", // Nothing to do (at maximum)
+                    "[Agent Loop] Waiting 10 ms before retrieving next work." // Nothing to do (at maximum)
+                );
+        }
+    }
+
+    @NotNull
+    private AgentWorkRetrievalScheduler createSchedulerForIterations(final ExponentialBackOff backoffStrategy, final int numIterations) {
+        return new AgentWorkRetrievalScheduler(controller, backoffStrategy, taskScheduler) {
+            int iterations;
+
+            @Override
+            void waitFor(long waitMillis) {
+                if (++iterations >= numIterations) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        };
+    }
+
+    private static ExponentialBackOff exponentialBackOffTwoToTen() {
+        ExponentialBackOff backoff = new ExponentialBackOff();
+        backoff.setInitialInterval(2);
+        backoff.setMultiplier(2);
+        backoff.setMaxInterval(10);
+        return backoff;
+    }
+
+    private static ExponentialBackOff instantBackOffTwoToTen() {
+        ExponentialBackOff backoff = new ExponentialBackOff();
+        backoff.setInitialInterval(2);
+        backoff.setMultiplier(10);
+        backoff.setMaxInterval(10);
+        return backoff;
+    }
+}

